### PR TITLE
Refactor shadow rule and modify shadow value generic type.

### PR DESCRIPTION
**Pull request by soulasuna**
## Refactor shadow rule and modify shadow value generic type.

For #11661.

Changes proposed in this pull request:
- Refactor the shadow rule add rule checker.
- Add shadow rule checker.
- Add shadow rule test.
- Modify shadow value generic type.

### CODE REVIEW

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/column/ColumnRegexMatchShadowAlgorithm.java [hunk 1]
```diff
@@ -81,11 +81,11 @@ private void checkOperation() {
     }
     
     @Override
-    public boolean isShadow(final Collection<String> shadowTableNames, final PreciseColumnShadowValue<String> shadowValue) {
+    public boolean isShadow(final Collection<String> shadowTableNames, final PreciseColumnShadowValue<Comparable<?>> shadowValue) {
         boolean containTable = shadowTableNames.contains(shadowValue.getLogicTableName());
```
Comment (taojintianxia) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/column/ColumnRegexMatchShadowAlgorithm.java:85]: why containTable is not isContainTable or why others start with `is`

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/rule/ShadowRule.java [hunk 2]
```diff
@@ -51,43 +56,95 @@
...
-    
-    private final Map<String, ShadowTableConfiguration> tables = new LinkedHashMap<>();
+    private final Map<String, String> shadowDataSourceMappings = new LinkedHashMap<>();
     
     private final Map<String, ShadowAlgorithm> shadowAlgorithms = new LinkedHashMap<>();
     
+    private final Map<String, ShadowTableRule> shadowTableRules = new LinkedHashMap<>();
+    
     public ShadowRule(final ShadowRuleConfiguration shadowRuleConfig) {
+        enable = shadowRuleConfig.isEnable();
```
Comment (taojintianxia) [inline shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/rule/ShadowRule.java:66]: upon the boolean variable style is `isXXXX`, but here enable is not `isEnable` . why not use same varible style ?

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/column/ColumnRegexMatchShadowAlgorithm.java (hunk 1 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/algorithm/shadow/column/ColumnRegexMatchShadowAlgorithm.java:85]: Thanks.
Method `isShadow()` in the algorithm is used to determine whether the sql contains shadow coloring. 
Shadow tables and shadow algorithms are separately configured and referenced by ID. 
If the user customizes the algorithm, he needs to get the configured shadow tables to customize his own judgment algorithm. 
So the beginning of `isContainTable` is not used.

**Review by soulasuna — COMMENTED** [review]
`isXXXX` is added by lombok annotation @Getter.

File: shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/rule/ShadowRule.java (hunk 2 above)
Comment (soulasuna) [reply shardingsphere-features/shardingsphere-shadow/shardingsphere-shadow-core/src/main/java/org/apache/shardingsphere/shadow/rule/ShadowRule.java:66]: isXXXX is added by lombok annotation @getter.

**Review by tristaZero — APPROVED** [review]

**Pull request merged**